pkg/lokifrontend/frontend/v2/frontend.go

```diff
@@ -144,19 +146,30 @@ func NewFrontend(cfg Config, ring ring.ReadRing, log log.Logger, reg prometheus.
 }
 
 func (f *Frontend) starting(ctx context.Context) error {
-	return errors.Wrap(services.StartAndAwaitRunning(ctx, f.schedulerWorkers), "failed to start frontend scheduler workers")
+	return errors.Wrap(services.StartAndAwaitRunning(context.Background(), f.schedulerWorkers), "failed to start frontend scheduler workers")
 }
 
-func (f *Frontend) stopping(_ error) error {
+func (f *Frontend) stopping(e error) error {
+	t := time.NewTicker(500 * time.Millisecond)
+	defer t.Stop()
+
+	start := time.Now()
+
+	for range t.C {
+		inflight := f.requests.count()
+		elapsed := time.Since(start)
+		if elapsed > f.cfg.GracefulShutdownTimeout || inflight <= 0 {
+			level.Debug(f.log).Log("msg", "no outstanding queries waiting to complete", "inflight", f.requests.count(), "elapsed", elapsed)
+			break
+		}
+		level.Debug(f.log).Log("msg", "outstanding queries waiting to complete", "inflight", inflight, "elapsed", elapsed)
+	}
+
 	return errors.Wrap(services.StopAndAwaitTerminated(context.Background(), f.schedulerWorkers), "failed to stop frontend scheduler workers")
 }
 
 // RoundTripGRPC round trips a proto (instead of a HTTP request).
 func (f *Frontend) RoundTripGRPC(ctx context.Context, req *httpgrpc.HTTPRequest) (*httpgrpc.HTTPResponse, error) {
-	if s := f.State(); s != services.Running {
```

Comment:
I think we should refuse new requests. There is `service.Stopping`.

Reply:
These are not "new" requests, these are sub-requests from a query request.

> During the time of the shutdown, the RoundTripGRCP function still needs to accept any round-tripping requests. This is, because the frontend round tripper is wrapped into a query range tripper ware, which splits incoming (HTTP) query requests into multiple (gRPC) query requests if needed.

Reply:
Not sure I understand the logic here. The query-frontend is the one responsible for splitting a HTTP request into multiple HTTP over GRPC requests that are then propagated to the queriers correct? So is this `RoundTripGRPC` method the one that handles the HTTP over GRPC responses from the queriers?

Reply:
Alright. Then ignore my request. How are we rejecting Incoming requests though?

Reply:
The frontend has a `/ready` endpoint that calls the `CheckReady(context.Context) error` handler. This endpoint returns `503` when the frontend service is not in `Running` state.

https://github.com/grafana/loki/blob/5a8e9900565545cdbf76437dc294945ce056a5a0/pkg/lokifrontend/frontend/v2/frontend.go#L301-L303

If running in K8s this will remove the instance from the query frontend service and therefore does not get any new requests from upstream (e.g. gateway).